### ISSUE TITLE
builtin `input` crashes during pip uninstall in virtual environment in win/git-bash when running `pip uninstall mypackage` crashes with: "input(): lost sys.stderr"

### ISSUE THREAD

**Issue by zackees** [python/cpython#99937]
# Bug report

This bug happens while in `git-bash` (and likely `cygwin`) while in a virtual environment. The work-around is to use `pip uninstall -y mypackage` to avoid the prompt which triggers the `input` builtin function.

See related pull request to put in a work-around for pip:

https://github.com/pypa/pip/pull/11630/files/4654a57a20ea36f3754e945bfe28f5e3581a0186

# Repro Steps

Repro steps:

  0. Start off in Windows 10 w/ git-bash
  1. `git clone https://github.com/zackees/keyvalue_sqlite`
  2. `cd keyvalue_sqlite`
  3. `python make_venv.py`
  4. `. activate.sh`
  5. `pip install -e .`
  6. ERROR STEP: `pip uninstall keyvalue_sqlite`,  outputs `"input(): lost sys.stderr"`

# Code affected

This is where `input` builtin is defined:
https://github.com/python/cpython/blob/main/Python/bltinmodule.c

```python
    /* Check that stdin/out/err are intact */
    if (fin == NULL || fin == Py_None) {
        PyErr_SetString(PyExc_RuntimeError,
                        "input(): lost sys.stdin");
        return NULL;
    }
    if (fout == NULL || fout == Py_None) {
        PyErr_SetString(PyExc_RuntimeError,
                        "input(): lost sys.stdout");
        return NULL;
    }
    if (ferr == NULL || ferr == Py_None) {
        PyErr_SetString(PyExc_RuntimeError,
                        "input(): lost sys.stderr");
        return NULL;
    }

    if (PySys_Audit("builtins.input", "O", prompt ? prompt : Py_None) < 0) {
        return NULL;
    }

    /* First of all, flush stderr */
    tmp = PyObject_CallMethodNoArgs(ferr, &_Py_ID(flush));
    if (tmp == NULL)
        PyErr_Clear();
    else
        Py_DECREF(tmp);
```

So it looks like `ferr == NULL or ferr == Py_None` is the problem.

# Your environment

- CPython versions tested on: 3.10.8, 3.9, 3.8 and probably others. I've seen this bug for a long time.
- Operating system and architecture:

![image](https://user-images.githubusercontent.com/6856673/205163166-5daf1e94-6240-4c2c-8b5e-78184f5abf83.png)

**Comment by ericvsmith** [comment]
Your make_venv.py script appears to install virtualenv into the global python environment. I say "appears", because I'm not going to run it on my system.

Can you reproduce whatever you're seeing without virtualenv, just using tools that ship with Python? And instead of a python program to demonstrate the problem, it would be better if you can show the individual steps on the command line.

**Comment by zackees** [comment]
> Can you reproduce whatever you're seeing without virtualenv

This will only reproduce while inside a virtualenv. It does not manifest while in the global python environment.

> And instead of a python program to demonstrate the problem, it would be better if you can show the individual steps on the command line.

Sure, try this:

  1. Start in git-bash
  2. `virtualenv -p python3 venv`
  3. `. venv/Scripts/activate`
  4. `pip install keyvalue_sqlite`
  5. ERROR STEP: `pip uninstall keyvalue_sqlite`

**Comment by ericvsmith** [comment]
Can you reproduce it with "python -m venv" instead of virtualenv? virtualenv is a third-party tool. If it's a problem with virtualenv, you should report it to https://github.com/pypa/virtualenv/issues

**Comment by zackees** [comment]
Yes, it reproduces using `python -m venv` too:

  1. Start off in git-bash
  2. `python -m venv venv`
  3. `. venv/Scripts/activate`
  4. `pip install keyvalue_sqlite`
  5. `pip uninstall keyvalue_sqlite`: RuntimeError: input(): lost sys.stderr

**Comment by zackees** [comment]
It looks like sys.stderr is None, even when `pip/__init__.py` is invoked.

This:
![image](https://user-images.githubusercontent.com/6856673/205182665-584b2248-ab66-419c-b22b-625b45521431.png)


Produces the output:
`sys.sterr is: None C:\Users\niteris\dev\transcribe-anything\venv\lib\site-packages\pip\__init__.py`

**Comment by zackees** [comment]
sys.stderr is None at the top level all the way to the builtin `runpy.py` file!

Here:
`C:\Users\niteris\AppData\Local\Programs\Python\Python310\lib\runpy.py`

What's strange is that launching any other program shows that it's NOT None!

$ pip uninstall transcribe-anything
sys.stderr is None C:\Users\niteris\AppData\Local\Programs\Python\Python310\lib\runpy.py
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "C:\Users\niteris\AppData\Local\Programs\Python\Python310\lib\runpy.py", line 15, in <module>
    traceback.print_stack()

**Comment by zackees** [comment]
So `runpy.py` is invoked from the `pip.exe` file

`sys.argv[0] = C:\Users\niteris\dev\transcribe-anything\venv\Scripts\pip.exe`

**Comment by zackees** [comment]
It's possibly related to this similar issue:

https://github.com/pypa/distlib/issues/168

**Comment by eryksun** [comment]
The linked [distlib issue](https://github.com/pypa/distlib/issues/168) is resolved in distlib version 0.3.5+, which is vendored by pip 22.2+. In Python 3.10.6, ensurepip bundles pip 22.2.1, and 3.10.7 and 3.10.8, it bundles pip 22.2.2. Please check the version of pip that you're using and the vendored version of distlib that pip is using.

**Comment by zackees** [comment]
It's confirmed that this is an issue with
https://github.com/pypa/distlib/issues/168

As distlib version 0.3.3 does NOT crash.

So I'm closing this bug.
